src/csr.hpp

```diff
@@ -89,88 +114,41 @@ class MatrixOperator
   /// The value type
   using value_type = T;
 
-  /// @brief Construct a CSR matrix for the Form a, with given boundary
-  /// conditions
-  /// @param a A finite element form
-  /// @param bcs Set of boundary conditions to be applied
+  /// @brief An on-device CSR Matrix
+  /// @param A CPU-based CSR Matrix
   MatrixOperator(
-      const dolfinx::fem::Form<T, T>& a,
-      std::vector<std::reference_wrapper<const dolfinx::fem::DirichletBC<T>>>
-          bcs)
-      : _comm(a.function_spaces()[0]->mesh()->comm())
+      dolfinx::la::MatrixCSR<T, std::vector<T>, std::vector<std::int32_t>,
+                             std::vector<std::int32_t>>& A)
   {
     dolfinx::common::Timer t0("~setup phase MatrixOperator");
 
-    if (a.rank() != 2)
-      throw std::runtime_error("Form should have rank be 2.");
-
-    auto V = a.function_spaces()[0];
-    dolfinx::la::SparsityPattern pattern
-        = dolfinx::fem::create_sparsity_pattern(a);
-    pattern.finalize();
-    _col_map = std::make_shared<const dolfinx::common::IndexMap>(
-        pattern.column_index_map());
-    _row_map = V->dofmap()->index_map;
-
-    _A = std::make_unique<
-        dolfinx::la::MatrixCSR<T, std::vector<T>, std::vector<std::int32_t>,
-                               std::vector<std::int32_t>>>(pattern);
-    dolfinx::fem::assemble_matrix(_A->mat_add_values(), a, bcs);
-    _A->scatter_rev();
-    dolfinx::fem::set_diagonal<T>(_A->mat_set_values(), *V, bcs, T(1.0));
+    // Copy to device
+    _A = std::make_unique<dolfinx::la::MatrixCSR<
+        T, thrust::device_vector<T>, thrust::device_vector<std::int32_t>,
+        thrust::device_vector<std::int32_t>>>(A);
```

Comment:
Should column index container have value type `std::int64_t`?

Reply:
The column index should be local, so will not need int64 (index map provides the mapping). The row pointer, could, perhaps get very big - over 2G, but in practice, the number of entries per GPU has been much less than that.

Reply:
We can easily exceed 200 non-zeros per row at higher order.

We won't always be chasing performance - sometimes we'll be testing for correctness.